server/src/test/java/org/elasticsearch/cluster/routing/allocation/decider/AllocationDecidersTests.java

```diff
@@ -20,255 +20,137 @@
 import org.elasticsearch.cluster.routing.ShardRouting;
 import org.elasticsearch.cluster.routing.UnassignedInfo;
 import org.elasticsearch.cluster.routing.allocation.RoutingAllocation;
+import org.elasticsearch.common.transport.TransportAddress;
 import org.elasticsearch.index.Index;
 import org.elasticsearch.index.shard.ShardId;
 import org.elasticsearch.test.ESTestCase;
-import org.hamcrest.Matcher;
-import org.hamcrest.Matchers;
 
-import java.util.Collection;
+import java.util.ArrayList;
 import java.util.List;
 import java.util.Optional;
 import java.util.Set;
+import java.util.function.BiFunction;
+import java.util.function.Supplier;
+import java.util.stream.Collector;
 
 import static org.hamcrest.Matchers.equalTo;
 
 public class AllocationDecidersTests extends ESTestCase {
 
-    public void testDebugMode() {
-        verifyDebugMode(RoutingAllocation.DebugMode.ON, Matchers.hasSize(1));
+    public void testCheckAllDecidersBeforeReturningYes() {
+        var expectedDecision = new Decision.Multi();
+        var allDecisions = generateDecision(() -> Decision.YES);
+
+        verifyDecidersCall(RoutingAllocation.DebugMode.OFF, allDecisions, allDecisions.size(), expectedDecision);
```

Comment:
I think we can indicate that the debug mode doesn't matter here:

```suggestion
        verifyDecidersCall(randomFrom(RoutingAllocation.DebugMode.values()), allDecisions, allDecisions.size(), expectedDecision);
```